tests/test_api_v1.py

```diff
@@ -2439,3 +2303,160 @@ def test_filter_requests(
+                        {"name": "d2", "replaces": None, "type": "gomod", "version": "2"},
+                    ],
+                },
+            ],
+            [
+                {"name": "d1", "type": "gomod", "version": "1"},
+                {"name": "d2", "replaces": None, "type": "gomod", "version": "2"},
+            ],
+        ],
+    ],
+)
+def test_fetch_request_packages_and_dependencies(
+    packages_data, expected_packages, expected_deps, app, db, client, auth_env, tmpdir
+):
+    """Test fetch a request with correct packages and dependencies read from packages data file."""
+    request = create_request_in_db(app, db, auth_env)
+
+    cachito_bundles_dir = str(tmpdir)
+    if packages_data is not None:
+        RequestBundleDir(request.id, root=cachito_bundles_dir).packages_data.write_text(
+            json.dumps(packages_data), encoding="utf-8"
+        )
+    app.config["CACHITO_BUNDLES_DIR"] = cachito_bundles_dir
+
+    rv = client.get(f"/api/v1/requests/{request.id}")
+
+    response_data = json.loads(rv.data)
+    assert response_data["packages"] == expected_packages
+    assert response_data["dependencies"] == expected_deps
+
+
+@pytest.mark.parametrize("verbose", [True, False])
+def test_fetch_requests_packages_and_dependencies(verbose, app, db, client, auth_env, tmpdir):
+    """Test packages and dependencies inside the fetched requests."""
+    request = create_request_in_db(app, db, auth_env)
+
+    packages_data = {
+        "packages": [
+            {
+                "name": "n2",
+                "type": "go-package",
+                "version": "v2",
+                "dependencies": [
+                    {"name": "d1", "type": "go-package", "version": "1"},
+                    {"name": "d2", "replaces": None, "type": "go-package", "version": "2"},
+                ],
+            },
+            {
+                "name": "n1",
+                "type": "gomod",
+                "version": "v1",
+                "dependencies": [
+                    {"name": "d1", "type": "gomod", "version": "1"},
+                    {"name": "d2", "replaces": None, "type": "gomod", "version": "2"},
+                ],
+            },
+            {
+                "name": "p1",
+                "type": "npm",
+                "version": "v2",
+                "dependencies": [{"name": "async", "type": "npm", "version": "1.2.0"}],
+            },
+            {
+                "name": "p2",
+                "type": "npm",
+                "version": "20210621",
+                "dependencies": [
+                    {"name": "async", "type": "npm", "version": "1.2.0"},
+                    {"name": "underscore", "type": "npm", "version": "1.13.0"},
+                ],
+            },
+        ],
+    }
+
+    expected_deps = [
+        {"name": "d1", "type": "go-package", "version": "1"},
+        {"name": "d2", "replaces": None, "type": "go-package", "version": "2"},
```

Comment:
Seeing this, I realized that Cachito currently always sets `"replaces": None` for dependencies that do not have replacements.

https://github.com/release-engineering/cachito/blob/ece43b86e31e7f858bd905cd58ed3887bd7e52d4/cachito/web/models.py#L477-L478
https://github.com/release-engineering/cachito/blob/ece43b86e31e7f858bd905cd58ed3887bd7e52d4/cachito/web/models.py#L903-L906

I'm not sure how much we need to worry about this detail of the API response, but if we want to be 100% backwards compatible, we need to do the same.

Reply:
@chmeliik 

Is this already covered in every `fetch_*_source` when writing the packages data? The resolved dependencies should have replaces, and they will be written into the package manager specific packages data file and then gathered into the `packages.json`.

Reply:
I don't think so. The dependencies returned by the `resolve_*` functions may or may not have the "replaces" key. In fact, I think only deps from `resolve_gomod` have the key.